src/OpenSSL/crypto.py

```diff
@@ -2042,54 +2042,153 @@ def export(self, cert, key, type=FILETYPE_PEM, days=100,
-            return True
-        return False
+        return _lib.PKCS7_type_is_enveloped(self._pkcs7)
 
     def type_is_signedAndEnveloped(self):
         """
-        Check if this NID_pkcs7_signedAndEnveloped object
+        Check if this is a NID_pkcs7_signedAndEnveloped object, which contains
+        encrypted data, encrypted decryption keys for each recipient, and
+        doubly encrypted message digests.
 
-        :returns: True if the PKCS7 is of type signedAndEnveloped
+        :returns: :const:`True` if the PKCS7 is of type
+            ``signedAndEnveloped``,:const:`False` otherwise.
+        :rtype: :class:`bool`
         """
-        if _lib.PKCS7_type_is_signedAndEnveloped(self._pkcs7):
-            return True
-        return False
+        return _lib.PKCS7_type_is_signedAndEnveloped(self._pkcs7)
 
     def type_is_data(self):
         """
-        Check if this NID_pkcs7_data object
+        Check if this is a NID_pkcs7_data object, which contains plain data.
 
-        :return: True if the PKCS7 is of type data
+        :return: :const:`True` if the PKCS7 is of type ``data``,
+            :const:`False` otherwise.
+        :rtype: :class:`bool`
         """
-        if _lib.PKCS7_type_is_data(self._pkcs7):
-            return True
-        return False
+        return _lib.PKCS7_type_is_data(self._pkcs7)
+
+    def type_is_digest(self):
+        """
+        Check if this is a NID_pkcs7_digested object, which contains
+        unencrypted data and a message digest.
+
+        :return: :const:`True` if the PKCS7 is of type ``digestedData``,
+            :const:`False` otherwise.
+        :rtype: :class:`bool`
+        """
+        return _lib.PKCS7_type_is_digest(self._pkcs7)
 
     def get_type_name(self):
         """
-        Returns the type name of the PKCS7 structure
+        Returns the type name of the PKCS7 structure.
 
-        :return: A string with the typename
+        :return: A string with the type name.
+        :rtype: :class:`str`
         """
         nid = _lib.OBJ_obj2nid(self._pkcs7.type)
         string_type = _lib.OBJ_nid2sn(nid)
         return _ffi.string(string_type)
 
+    def get_certificates(self):
+        """
+        Returns all certificates for the PKCS7 structure, if present. Only
+        objects of type ``signedData`` or ``signedAndEnvelopedData`` can embed
+        certificates.
+
+        :return: The certificates in the PKCS7, or :const:`None` if
+            there are none.
+        :rtype: :class:`tuple` of :class:`X509` or :const:`None`
+        """
+        certs = _ffi.NULL
+        if self.type_is_signed():
+            certs = self._pkcs7.d.sign.cert
+        elif self.type_is_signedAndEnveloped():
+            certs = self._pkcs7.d.signed_and_enveloped.cert
+
+        pycerts = []
+        for i in range(_lib.sk_X509_num(certs)):
+            pycert = X509.__new__(X509)
+            pycert._x509 = _lib.sk_X509_value(certs, i)
```

Comment:
Deleting original pkcs7 object will break extracted cert, because `sk_X509_value` returns internal pointer.
You should duplicate certificate with `_lib.X509_dup` and do not not forget to free them `pycert._x509 = _ffi.gc(_lib.X509_dup(_lib.sk_X509_value(sk_stack, i)), _lib.X509_free)`